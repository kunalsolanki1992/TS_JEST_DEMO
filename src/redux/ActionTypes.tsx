export default {
    LOGIN: 'LOGIN',
    FETCH_TODO_REQUEST: 'FETCH_TODO_REQUEST',
    FETCH_TODO_SUCCESS: 'FETCH_TODO_SUCCESS',
    FETCH_TODO_FAILURE: 'FETCH_TODO_FAILURE'
}