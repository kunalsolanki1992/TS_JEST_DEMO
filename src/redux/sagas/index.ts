import { all, fork } from "redux-saga/effects";

import todoSaga from "./Auth";

export function* rootSaga() {
    yield all([fork(todoSaga)])
}
