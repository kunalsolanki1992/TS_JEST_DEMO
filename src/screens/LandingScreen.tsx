import React, { useState, useReducer, useEffect } from "react";
import { View, Text, StyleSheet, Dimensions, Image } from 'react-native';
import GetLocation, {Location } from 'react-native-get-location';

const screenWidth = Dimensions.get('screen').width;

export const LandingScreen = () => {
    const [errorMsg, setErrorMsg] = useState("");
    const [address, setAddress] = useState();
    const [displayAddress, setDisplayAddress] = useState("")

    useEffect(() => {
        (() => {
            // let { status } = await GetLocation.getCurrentPosition
            GetLocation.getCurrentPosition({
                enableHighAccuracy: true,
                timeout: 15000,
            })
            .then(location => {
                console.log(location);
            })
            .catch(error => {
                const { code, message } = error;
                console.warn(code, message);
            })
        })()
    }, [])


    return (
        <View style={styles.container}>
            <View style={styles.navigation} />

            <View style={styles.body}>
                <Image
                    source={{ uri: 'https://elasq.com/wp-content/uploads/2021/08/truck4.png' }}
                    style={styles.deliveryIcon}
                    resizeMode={'center'}
                />

                <View style={styles.addressContainer}>
                    <Text style={styles.addressTitle}>Your Delivery Address</Text>
                </View>

                <Text style={styles.addressText}>Waiting for current location</Text>
            </View>

            <View style={styles.footer}>
                <Text>Footer</Text>
            </View>
        </View>
    )
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        // backgroundColor: 'green'
        backgroundColor: 'rgba(242,242,242,1)'
    },

    navigation: {
        flex: 2,
        // backgroundColor: 'red'
    },

    body: {
        flex: 9,
        justifyContent: 'center',
        alignItems: 'center',
        // backgroundColor: 'yellow'
    },

    deliveryIcon: {
        width: 120,
        height: 120
    },

    addressContainer: {
        width: screenWidth - 100,
        borderBottomColor: 'red',
        borderBottomWidth: 0.5,
        padding: 5,
        marginBottom: 10,
        alignItems: 'center'
    },

    addressTitle: {
        fontSize: 24,
        fontWeight: '700',
        color: '#7d7d7d'
    },

    addressText: {
        fontSize: 20,
        fontWeight: '200',
        color: '#4f4f4f'
    },

    footer: {
        flex: 1,
        // backgroundColor: 'cyan'
    }
})