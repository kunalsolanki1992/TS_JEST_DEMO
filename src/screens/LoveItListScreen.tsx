// LOVE IT LIST BOTTOM TAB
import React from "react";
import { View, Text, StyleSheet, Dimensions, Image } from 'react-native';
import { useSelector } from 'react-redux';

// component imports
import ParentContainer from "../components/ParentContainer";

export const LoveItListScreen = (props: any) => {
    const userData = useSelector((state: any) => state.user.user);

    return (
        <ParentContainer {...props}>
            <View style={styles.body}>
                <Text>{!!userData && !!userData.email ? `Hello ${userData.email}, Love-it List` : 'Love-it List Screen'}</Text>
            </View>
        </ParentContainer>
    )
};

const styles = StyleSheet.create({
    body: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
})