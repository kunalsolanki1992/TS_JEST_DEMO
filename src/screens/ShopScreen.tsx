// SHOP BOTTOM TAB
import React from "react";
import { View, Text, StyleSheet, Dimensions, Image } from 'react-native';
import { useSelector } from 'react-redux';

// component imports
import ParentContainer from "../components/ParentContainer";

export const ShopScreen = (props: any) => {
    const userData = useSelector((state: any) => state.user.user);

    console.log("USER DATA == ", userData);

    return (
        <ParentContainer {...props}>
            <View style={styles.body}>
                <Text>{!!userData && !!userData.email ? `Hello ${userData.email}, Shop Screen` : 'Shop Screen'}</Text>
            </View>
        </ParentContainer>
    )
};

const styles = StyleSheet.create({
    body: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        // backgroundColor: 'yellow'
    },
})