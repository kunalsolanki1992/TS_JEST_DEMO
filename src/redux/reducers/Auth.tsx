import ActionTypes from "../ActionTypes";
import { TodoState, TodoActions } from './types';

const initialState: TodoState = {
    user: {},
    pending: false,
    todos: [],
    error: null,
}

const LoginReducer = (state = initialState, action: any) => {
    switch (action.type) {
        case ActionTypes.LOGIN:
            return {
                ...state,
                user: action.payload
            }

        case ActionTypes.FETCH_TODO_REQUEST:
            return {
                ...state,
                pending: true
            }

        case ActionTypes.FETCH_TODO_SUCCESS:
            return {
                ...state,
                pending: false,
                todos: action.payload.todos,
                error: null
            }

        case ActionTypes.FETCH_TODO_SUCCESS:
            return {
                ...state,
                pending: false,
                todos: [],
                error: action.payload.error
            }

        default:
            return initialState;
    }
}

export default LoginReducer;