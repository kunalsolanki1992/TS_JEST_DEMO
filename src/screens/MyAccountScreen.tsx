// MY ACCOUNT BOTTOM TAB
import React from "react";
import { View, Text, StyleSheet, Dimensions, Image, TouchableOpacity, TextInput, Keyboard, Alert, Button } from 'react-native';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { useDispatch, useSelector } from 'react-redux';

// redux imports
import { login } from "../redux/actions/Auth";

// constant imports
import { DEVICE_WIDTH } from "../constants/dimens";

// component imports
import ParentContainer from "../components/ParentContainer";

const COMP_WIDTH = DEVICE_WIDTH / 5;

export const MyAccountScreen: React.FC = (props: any) => {
    const [stateValue, setStateValue] = React.useState(true);
    const [email, setEmail] = React.useState<string>("");
    const [password, setPassword] = React.useState<string>("");

    const userData = useSelector((state: any) => state.user.user);

    const dispatch = useDispatch();

    const isValid = () => {
        const emailRegex = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

        if (email === '') {
            Alert.alert("Please enter Email!");
            return false;
        }

        if (password === '') {
            Alert.alert("Please enter password!")
            return false;
        }

        if (!email.match(emailRegex)) {
            Alert.alert("Please enter valid email!");
            return false
        }

        return true;
    }

    const signIn = () => {
        if (isValid()) {
            dispatch(login({ email }))
        }
    }

    const signOff = () => { dispatch(login({})) }

    return (
        <ParentContainer {...props}>
            <View style={styles.container}>
                {!!userData && userData.email === undefined && (<View style={styles.loginContainer}>
                    <TextInput
                        style={styles.inputStyle}
                        value={email}
                        placeholder={"E-mail"}
                        onChangeText={(text) => setEmail(text)}
                        keyboardType={'email-address'}
                        onSubmitEditing={() => Keyboard.dismiss()}
                    />

                    <TextInput
                        style={styles.inputStyle}
                        value={password}
                        placeholder={"Password"}
                        onChangeText={(text) => setPassword(text)}
                        // keyboardType={'phone-pad'}
                        onSubmitEditing={() => Keyboard.dismiss()}
                    />

                    {/* <TouchableOpacity style={styles.btnToggleTCContainer} onPress={signIn}>
                    <Ionicons name={'log-in'} size={24} style={{ color: 'white' }} />

                    <Text style={styles.btnText} >{'Login'}</Text>
                </TouchableOpacity> */}

                    <Button
                        title="Login"
                        onPress={signIn}
                    />
                </View>)}

                {!!userData && !!userData.email && (
                    <TouchableOpacity style={styles.btnToggleTCContainer} onPress={signOff}>
                        <Ionicons name={'log-out'} size={24} style={{ color: 'white' }} />

                        <Text style={styles.btnText} >{'Logout'}</Text>
                    </TouchableOpacity>
                )}

                {/* <View style={styles.toggleUITestCaseContainer}>
                <TouchableOpacity style={styles.btnToggleTCContainer} onPress={() => { setStateValue(!stateValue) }} >
                    <Ionicons name={'ios-build'} size={24} style={{ color: 'white' }} />

                    <Text style={styles.btnText} >{stateValue ? 'Pass' : 'Fail'}</Text>
                </TouchableOpacity>
            </View> */}
            </View>
        </ParentContainer>
    )
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        // justifyContent: 'space-evenly'
        // backgroundColor: 'green'
    },

    inputStyle: {
        width: '100%',
        height: 20,
        borderBottomColor: '#aaa',
        borderBottomWidth: 1,
        minHeight: 40,
        marginBottom: 8
    },

    loginContainer: {
        width: DEVICE_WIDTH,
        padding: 20,
        minHeight: 1,
        // backgroundColor: 'pink',
        alignItems: 'center'
    },

    toggleUITestCaseContainer: {
        width: DEVICE_WIDTH,
        height: COMP_WIDTH,
        // backgroundColor: 'yellow'
    },

    btnToggleTCContainer: {
        flexDirection: 'row',
        padding: 10,
        width: 100,
        borderRadius: 4,
        backgroundColor: 'orange',
        alignSelf: 'center',
        alignItems: 'center',
        justifyContent: 'space-evenly',
        marginTop: 10
    },

    btnText: {
        fontSize: 18,
        color: 'white'
    },
})