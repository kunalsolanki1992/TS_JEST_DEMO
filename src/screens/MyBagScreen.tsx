// MY BAG BOTTOM TAB
import React from "react";
import { View, Text, StyleSheet, Dimensions, Image, Button, ActivityIndicator, FlatList } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';

// redux imports
import { fetchTodoRequest } from '../redux/actions/Auth';

// component imports
import ParentContainer from "../components/ParentContainer";

// constant imports
import { DEVICE_WIDTH, HORIZONTAL_DIMENS } from '../constants/dimens';

export const MyBagScreen = (props: any) => {
    const userData = useSelector((state: any) => state.user.user);
    const pending = useSelector((state: any) => state.user.pending);
    const todos = useSelector((state: any) => state.user.todos);
    const error = useSelector((state: any) => state.user.error);

    const dispatch = useDispatch();

    const callTodos = () => { dispatch(fetchTodoRequest()) }

    return (
        <ParentContainer {...props}>
            <View style={styles.body}>
                <Text>{!!userData && !!userData.email ? `Hello ${userData.email}, My Bag` : 'My Bag Screen'}</Text>
            </View>

            <Button
                title="Click Me"
                onPress={callTodos}
            // style={{ backgroundColor: 'blue' }}
            />

            {pending && (
                <ActivityIndicator size={'small'} />
            )}

            <FlatList
                data={todos}
                renderItem={(item) => (
                    <View style={styles.listItem}>
                        <Text>{item.item.title}</Text>
                    </View>
                )}
            />
        </ParentContainer>
    )
};


const styles = StyleSheet.create({
    body: {
        // flex: 9,
        justifyContent: 'center',
        alignItems: 'center',
        margin: HORIZONTAL_DIMENS._20
        // backgroundColor: 'yellow'
    },

    listItem: {
        width: DEVICE_WIDTH,
        padding: HORIZONTAL_DIMENS._15,
        backgroundColor: 'grey'
    }
})