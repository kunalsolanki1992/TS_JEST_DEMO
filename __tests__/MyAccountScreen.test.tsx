import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { PersistGate } from 'redux-persist/lib/integration/react';
import { fireEvent, render } from '@testing-library/react-native';
import '@testing-library/jest-dom';

import { store, persistor } from '../src/redux/reducers';
import { MyAccountScreen } from '../src/screens/MyAccountScreen';

jest.mock("@react-native-async-storage/async-storage", () => ({
    setItem: jest.fn(() => Promise.resolve()),
    getItem: jest.fn(() => Promise.resolve()),
}))

jest.mock('redux-persist', () => {
    const real = jest.requireActual('redux-persist');
    return {
        ...real,
        persistReducer: jest
            .fn()
            .mockImplementation((config, reducers) => reducers),
    };
});

// tried to mock useState hook

// FAILED ATTEMPT 4
// jest.mock('react', () => {
//     const tempReact = jest.requireActual('react');
//     return {
//         ...jest.requireActual('react'),
//         useState: jest.fn()
//     }
// })

// const tempUseState: jest.Mock<typeOf useState> = useState as never;

// const setValue = jest.fn();
// tempUseState
//     .mockImplementation(() => ['value', setValue]);

// FAILED ATTEMPT 3
// const stateSetter = jest.fn()
// jest
// .spyOn(React, 'useState')
// //Simulate that mode state value was set to 'new mode value'
// .mockImplementation(stateValue => [stateValue='new mode value', stateSetter])

// FAILED ATTEMPT 1
// const handleClick = jest.spyOn(React, "useState");
// handleClick.mockImplementation((size) => [size, changeSize]);

// FAILED ATTEMPT 2
// jest.mock('react', () => ({
//     ...jest.requireActual('react'),
//     useState: jest.fn()
// }));

// const useStateMock: jest.Mock<typeof useState> = useState as never;

// defined new state here.
// const setValue = jest.fn();
// useStateMock
//     .mockImplementation(() => ['value', setValue]);

it('My Account screen renders correctly', () => {
    const tree = renderer
        .create(
            <Provider store={store}>
                {/* <PersistGate persistor={persistor}> */}
                <MyAccountScreen />
                {/* </PersistGate> */}
            </Provider>
        )
        .toJSON();

    // console.log(tree);
    // console.log(setValue);

    expect(tree).toMatchSnapshot();
});

describe('Other tests in My Account Screen', () => {
    let loginButton;
    // let passwordText;

    const setEmailMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setEmailMock];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);

    const { getByText, getByPlaceholderText } = render(<Provider store={store}>
        <MyAccountScreen />
    </Provider>)

    // passwordText = getByText("123456");
    let emailText = getByPlaceholderText('E-mail');
    let passwordText = getByPlaceholderText('Password');
    loginButton = getByText('Login');

    // fireEvent.click('loginButton');
    // fireEvent.changeText(passwordText);
    fireEvent.press(loginButton);

    // expect(useStateMock).toHaveBeenCalledTimes(3);
    
    // emailText.props.value = 'kunal.solanki@capgemini.com';
    // passwordText.props.value = '123456';
    
    const newEmail = 'kunal.solanki@capgemini.com';
    const newPassword = '123456';

    fireEvent.changeText(emailText, newEmail);
    fireEvent.changeText(passwordText, newPassword);
    
    // console.log("Login Button == ", loginButton);
    console.log("Email TextInput == ", emailText.props.value);
    // console.log("Password TextInput == ", passwordText);
    
    // expect(emailText.props.value).toMatch(/^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/)
    expect(passwordText.props.value).toMatch("");
    expect(setEmailMock).toHaveBeenCalledTimes(2);
    // expect(loginButton).toHaveBeenCalledTimes(1);
});