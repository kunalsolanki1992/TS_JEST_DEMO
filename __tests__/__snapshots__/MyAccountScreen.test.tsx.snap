// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Account screen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "minHeight": 1,
          "padding": 20,
          "width": 750,
        }
      }
    >
      <TextInput
        keyboardType="email-address"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="E-mail"
        style={
          Object {
            "borderBottomColor": "#aaa",
            "borderBottomWidth": 1,
            "height": 20,
            "marginBottom": 8,
            "minHeight": 40,
            "width": "100%",
          }
        }
        value=""
      />
      <TextInput
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Password"
        style={
          Object {
            "borderBottomColor": "#aaa",
            "borderBottomWidth": 1,
            "height": 20,
            "marginBottom": 8,
            "minHeight": 40,
            "width": "100%",
          }
        }
        value=""
      />
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
